### PR TITLE
build: make Zephyr stdint convention enforcing configurable

### PR DESCRIPTION
**Pull request by npitre**
This is my take on addressing #46032.

There are  few cases where removing the Zephyr stdint enforcement is desirable:

- linking against binary C++ libs with incompatible type mangling

- linking against system provided headers i.e. native_posix

- compiling with legacy code that assumes a different convention

So let's create a Kconfig symbol for it. This is IMHO a good compromize
compared to using the `%"PRId32"` abomination everywhere otherwise.

### CODE REVIEW

File: include/zephyr/toolchain/gcc.h [hunk 1]
```diff
@@ -213,7 +213,20 @@ do {                                                                    \
...
+/*
+ * The Zephyr stdint convention enforces int32_t = int, int64_t = long long,
+ * and intptr_t = long so that short string format length modifiers can be
+ * used universally across ILP32 and LP64 architectures. Without that it
+ * is possible for ILP32 toolchains to have int32_t = long and intptr_t = int
+ * clashing with the Zephyr convention and generating pointless warnings
+ * as they're still the same size. Inhibit the format argument type
+ * validation in that case and let the other configs do it.
+ */
+#define __printf_like(f, a)
```
Comment (keith-packard) [inline include/zephyr/toolchain/gcc.h:228]: This will only catch internal Zephyr APIs using this macro; places using stdio APIs will still generate warnings.

File: Kconfig.zephyr [hunk 2]
```diff
@@ -608,6 +608,22 @@ config WARN_EXPERIMENTAL
 	  Print a warning when the Kconfig tree is parsed if any experimental
 	  features are enabled.
 
+config ENFORCE_ZEPHYR_STDINT
+	bool
+	prompt "Enforce Zephyr convention for stdint"
+	depends on !ARCH_POSIX
```
Comment (keith-packard) [inline Kconfig.zephyr:614]: If one goal is to avoid issues with C++ libraries, should this also be disabled when building C++ code?

**Review by keith-packard — COMMENTED** [review]
Overall, while I appreciate the goal of making code more readable, it's hard to get excited about overriding the toolchain for this. You've already identified one potential issue (C++ ABI), which makes me concerned that there will be additional issues popping up in the future. I'd prefer to accept the pain of %"PRId32" as that ensures compatibility with whatever the toolchain does.

**Comment by npitre** [comment]
This Zephyr stdint enforcement has been in place for 3 years now.

It works with the Zephyr provided stdio API just fine.

It even works with the whole of Zephyr built Newlib just fine.

It also works with whatever C++ code you may build alongside Zephyr.

It is a problem with C++ _only_ if you have binary-only libs _and_ it was
compiled using a toolchain configured to map int32_t to a long or some other
random choices like that (and yes that exists because of course the standard
allows it so why not?).

So let's stop making a strawman out of this. Either the issues are hypothetical
as in "we want Zephyr to be build using as many different compilers as we can,"
or they are real and this proposal covers each of those few ones that have been encountered in those last 3 years.

**Review by andyross — DISMISSED** [review]
(I just reviewed this, but it didn't appear?  Trying again, if I can remember how I phrased it...)

This one gets my vote.  It's a wart, but it's an obscure hidden wart that only platforms with odd configurations need to worry about, and the cost for them is low and tolerable ("set this kconfig and walk away").

The PRI stuff is a wart we'll all carry around on our faces forever, and the cost is actually quite high.  No one understands it or why it's needed, it's not routine syntax in any other major C project[1], and so submitters are going to be fighting CI failures on newlib, etc... again, and again, and again, and again.

[1] And in particular people coming from Linux backgrounds are going to absolutely hate it.

**Review by stephanosio — COMMENTED** [review]
I am not opposed to this as a temporary workaround, but I do believe https://github.com/zephyrproject-rtos/zephyr/issues/46032 is an inevitable direction for the reasons well explored in the issue: https://github.com/zephyrproject-rtos/zephyr/issues/46032.

For one, this is not a standard approach: `int32_t`, a standard fixed-width integer type, ought to be printed via `PRI32d`, the corresponding standard format specifier. Printing `int32_t` with `%d` is not standard conformant and can be said to be an **incorrect** usage of such. Adding a hack to make that work is not a sustainable approach for the reasons extensively described in https://github.com/zephyrproject-rtos/zephyr/issues/46032#issuecomment-1139314950.

If we are going to implement a non-standard method of printing the C99 fixed-width integer types, then do it; do not change the **standard functions** to make them (and rely on them to) behave correctly in non-standard way.

**Comment by stephanosio** [comment]
> The PRI stuff is a wart we'll all carry around on our faces forever, and the cost is actually quite high. No one understands it or why it's needed, it's not routine syntax in any other major C project[1], and so submitters are going to be fighting CI failures on newlib, etc... again, and again, and again, and again.

I have already suggested a workaround for this in https://github.com/zephyrproject-rtos/zephyr/issues/46032#issuecomment-1139228035.

**Comment by npitre** [comment]
> If we are going to implement a non-standard method of printing the
> C99 fixed-width integer types, then do it; do not change the **standard
> functions** to make them (and rely on them to) behave correctly in
> non-standard way.

Let me dare asking: why not?

Yes this is a "hack". And it may well be "temporary". Temporary is the
fate of all things in the universe after all.

If the cost of carrying this "hack" becomes greater than the fully
standard-confirming alternative, or gets in the way of people using
that standard, then yes a better approach would require consideration.
But at the moment I think the overall benefit vs cost is still in favor
of this standard-bending approach.

**Comment by keith-packard** [comment]
I think this means that Zephyr requires sizeof(int) == sizeof(int32_t) and sizeof(long long) == sizeof(int64_t). Is there somewhere in the system which checks for this? If not, adding a test would at least catch systems which violate this assumption.

**Comment by npitre** [comment]
> I think this means that Zephyr requires sizeof(int) == sizeof(int32_t)
> and sizeof(long long) == sizeof(int64_t).

Exact.

> Is there somewhere in the system which checks for this? If not, adding
> a test would at least catch systems which violate this assumption.

Very good point. This is already tested for in
include/zephyr/toolchain/zephyr_stdint.h but let's make sure this
remains true when that is disabled.

**Comment by npitre** [comment]
> If not, adding a test would at least catch systems which violate
> this assumption.

Patch refreshed.

**Review by andyross — APPROVED** [review]
> Patch refreshed.

Review refreshed.  Trying desperately not to get involved with the bikeshed except to say that this is absolutely my preferred tactic.

**Review by keith-packard — APPROVED** [review]
I just noticed the BUILD_ASSERT for the relevant types. This looks like it will work to me.

**Comment by stephanosio** [comment]
> Let me dare asking: why not?

I have already explained why not in https://github.com/zephyrproject-rtos/zephyr/issues/46032#issuecomment-1139314950 -- hypothetical or not, and also not to mention how it is a very real problem for C++, the approach taken in the `zephyr_stdint.h` is an explicit violation of the C standard and we are doomed to rework everything at some point in the future when there exist irreconcilable differences (e.g. 16-bit int, 128-bit long long, a toolchain-bundled C library that does not care about the non-standard `__INT32_TYPE__` and its friends).

To reiterate what has been said in the aforementioned comment, `zephyr_stdint.h` is not a sustainable solution let alone a portable one; as an operating system project, we need to seek both a sustainable and a portable solution.

If you do not like something from the C standard, you have the option of: a) throwing it away altogether, b) using it as-is no matter how horrible it may be. You simply cannot pick one thing you like and throw away the rest that comes with/is required by it -- if you do that, you are going to mess up a lot of things.

**TL;DR: If you want to use `int32_t` and its friends in the `printf`, use `PRId32` and its friends too. Do not use `int32_t` in the `printf` and drop the accompanying `PRId32` because that is going to create a big mess.**

I too agree with you that the `PRIxN` macros in the `printf` are horrible and we do not want to live with this wart in the Zephyr codebase. **So let us drop it all. Let us switch over to using the `printk` and the Zephyr-native cbprintf API** with the alternative C99 fixed-width integer type format specifiers like `%I32d` for `int32_t` as suggested in https://github.com/zephyrproject-rtos/zephyr/issues/46032#issuecomment-1139228035.

**Review by stephanosio — APPROVED** [review]
Although not ideal, this is an improvement over the status quo and a sufficiently reasonable temporary workaround.

File: Kconfig.zephyr [hunk 3]
```diff
@@ -608,6 +608,22 @@ config WARN_EXPERIMENTAL
 	  Print a warning when the Kconfig tree is parsed if any experimental
 	  features are enabled.
 
+config ENFORCE_ZEPHYR_STDINT
+	bool
+	prompt "Enforce Zephyr convention for stdint"
```
Comment (tejlmand) [inline Kconfig.zephyr:613]: late to the game, but should this really have a prompt ?

Wouldn't it be better to have this promptless so for example a pre-built library can makes a `Kconfig` file that will default this symbol to `n` when said library is selected and linked ?

I believe it would be for such libraries to know whether or not this should be enforced and not the enduser.

**Review by tejlmand — COMMENTED** [review]
@keith-packard  a little late to the game, but I wonder if this symbol should be promptless.

Why should the end-user know about this, as I see it, the build system ought to know when it has been configured to link a pre-built library that requires this setting to be `n`.
